plugins/trace_api_plugin/compressed_file.cpp

```diff
@@ -0,0 +1,311 @@
+   std::vector<uint8_t> read_buffer = std::vector<uint8_t>(read_buffer_size);
+   size_t remaining_read_buffer = 0;
+   bool initialized = false;
+   size_t file_size = 0;
+};
+
+compressed_file::compressed_file( fc::path file_path )
+:file_path(std::move(file_path))
+,file_ptr(nullptr)
+,impl(std::make_unique<compressed_file_impl>())
+{
+   impl->file_size = fc::file_size(file_path);
+}
+
+compressed_file::~compressed_file()
+{}
+
+void compressed_file::seek( long loc ) {
+   impl->seek(loc, *file_ptr);
+
+}
+
+void compressed_file::read( char* d, size_t n ) {
+   impl->read(d, n, *file_ptr);
+}
+
+// these are defaulted now that the opaque impl type is known
+//
+compressed_file::compressed_file( compressed_file&& ) = default;
+compressed_file& compressed_file::operator= ( compressed_file&& ) = default;
+
+
+/**
+ * Create a compressed file that looks like this:
+ *
+ * /====================\ file offset 0
+ * |                    |
+ * |  Compressed Data   |
+ * |  with seek points  |
+ * |                    |
+ * |--------------------|  file offset END - 2 - (16 * seek point count)
+ * |                    |
+ * |  mapping of        |
+ * |    orig offset to  |
+ * |    seek pt offset  |
+ * |                    |
+ * |--------------------|  file offset END - 2
+ * |  seek pt count     |
+ * \====================/  file offset END
+ *
+ * Where a "seek point" is a point in the compressed data stream where
+ * the decompressor can start reading from having not read any of the prior data
+ * seek points should be traversable by a decompressor so that reads which span
+ * seek points do not have to be aware of them
+ *
+ * In zlib this is created by doing a complete flush of the stream
+ *
+ * @param input_path
+ * @param output_path
+ * @param seek_point_count
+ * @return
+ */
+bool compressed_file::process( const fc::path& input_path, const fc::path& output_path, uint16_t seek_point_count ) {
+   if (!fc::exists(input_path)) {
+      throw std::ios_base::failure(std::string("Attempting to create compressed_file from file that does not exist: ") + input_path.generic_string());
+   }
+   std::vector<seek_point_entry> seek_point_map(seek_point_count);
+
+   size_t input_size = fc::file_size(input_path);
+   auto ideal_seek_stride = input_size / seek_point_map.size();
+
+   fc::cfile input_file;
+   input_file.set_file_path(input_path);
+   input_file.open("rb");
+
+   fc::cfile output_file;
+   output_file.set_file_path(output_path);
+   output_file.open("wb");
+
+   z_stream strm;
+   strm.zalloc = Z_NULL;
+   strm.zfree = Z_NULL;
+   strm.opaque = Z_NULL;
+
+   if (deflateInit2(&strm, Z_BEST_COMPRESSION, Z_DEFLATED, -15, 8, Z_DEFAULT_STRATEGY) != Z_OK) {
+      return false;
+   }
+
+   constexpr size_t buffer_size = 64*1024;
+   auto input_buffer = std::vector<uint8_t>(buffer_size);
+   auto output_buffer = std::vector<uint8_t>(buffer_size);
+
+   auto bytes_remaining_before_sync = ideal_seek_stride;
+   int next_sync_point = 0;
```

Comment:
verify/assert ideal_seek_stride is less than buffer_size, std::min check below can drop checking buffer_size

Reply:
I was confused on this, and thus resolving it now that I am not confused.